contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapV2Reserve.sol

```diff
@@ -0,0 +1,324 @@
+pragma solidity 0.6.6;
+
+import "@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol";
+import "@uniswap/v2-core/contracts/interfaces/IUniswapV2Pair.sol";
+import "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router01.sol";
+
+import "../../../IKyberReserve.sol";
+import "../../../IERC20.sol";
+import "../../../utils/Withdrawable3.sol";
+import "../../../utils/Utils5.sol";
+import "../../../utils/zeppelin/SafeERC20.sol";
+
+contract KyberUniswapV2Reserve is IKyberReserve, Withdrawable3, Utils5 {
+    using SafeERC20 for IERC20;
+
+    uint256 public constant DEFAULT_FEE_BPS = 0;
+    uint256 public constant DEADLINE = 2**255;
+
+    address public immutable kyberNetwork;
+    // fee deducted for each trade
+    uint256 public feeBps = DEFAULT_FEE_BPS;
+
+    bool public tradeEnabled = true;
+
+    IUniswapV2Router01 public immutable uniswapRouter;
+    IUniswapV2Factory public immutable uniswapFactory;
+    address public immutable weth;
+
+    mapping(IERC20 => bool) public tokenListed;
+    mapping(IERC20 => address[][]) public e2tSwapPaths;
+    mapping(IERC20 => address[][]) public t2eSwapPaths;
+
+    event TradeExecute(
+        address indexed sender,
+        IERC20 indexed srcToken,
+        uint256 srcAmount,
+        IERC20 indexed destToken,
+        uint256 destAmount,
+        address destAddress
+    );
+
+    event TokenListed(IERC20 indexed token, bool add);
+
+    event TokenPathAdded(IERC20 indexed token, address[] path, bool isEthToToken, bool add);
+
+    event TradeEnabled(bool enable);
+
+    event FeeUpdated(uint256 feeBps);
+
+    event EtherReceival(address indexed sender, uint256 amount);
+
+    constructor(
+        IUniswapV2Router01 _uniswapRouter,
+        address _weth,
+        address _admin,
+        address _kyberNetwork
+    ) public Withdrawable3(_admin) {
+        require(address(_uniswapRouter) != address(0), "uniswapRouter 0");
+        require(_weth != address(0), "weth 0");
+        require(_kyberNetwork != address(0), "kyberNetwork 0");
+
+        uniswapRouter = _uniswapRouter;
+        uniswapFactory = IUniswapV2Factory(_uniswapRouter.factory());
+        weth = _weth;
+        kyberNetwork = _kyberNetwork;
+    }
+
+    receive() external payable {
+        emit EtherReceival(msg.sender, msg.value);
+    }
+
+    /**
+      conversionRate: expected conversion rate should be >= this value.
+     */
+    function trade(
+        IERC20 srcToken,
+        uint256 srcAmount,
+        IERC20 destToken,
+        address payable destAddress,
+        uint256 conversionRate,
+        bool /* validate */
+    ) external override payable returns (bool) {
+        require(tradeEnabled, "trade is disabled");
+        require(msg.sender == kyberNetwork, "only kyberNetwork");
+        require(isValidTokens(srcToken, destToken), "token is not listed");
```

Comment:
maybe a better message.
it will fail if one side not eth.
or one side not listed.

Maybe "only use eth and listed token"

Reply:
change as request